eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/MultiPartHandler.scala

```diff
@@ -100,41 +125,41 @@ class MultiPartHandler(nodeParams: NodeParams, db: IncomingPaymentsDb, commandBu
+        log.warning(s"payment with paidAmount=${parts.map(_.amount).sum} failed ($failure)")
         pendingPayments.get(paymentHash).foreach { case (_, handler: ActorRef) => handler ! PoisonPill }
-        parts.foreach(p => commandBuffer ! CommandBuffer.CommandSend(p.payment.fromChannelId, CMD_FAIL_HTLC(p.htlcId, Right(failure), commit = true)))
+        doFail(e)
         pendingPayments = pendingPayments - paymentHash
       }
 
-    case MultiPartPaymentFSM.MultiPartHtlcSucceeded(paymentHash, parts) if doHandle(paymentHash) =>
+    case e@MultiPartPaymentFSM.MultiPartPaymentSucceeded(paymentHash, parts) if doHandle(paymentHash) =>
       Logs.withMdc(log)(Logs.mdc(paymentHash_opt = Some(paymentHash))) {
-        val received = PaymentReceived(paymentHash, parts.map(_.payment))
-        log.info(s"received complete payment for amount=${received.amount}")
-        // The first thing we do is store the payment. This allows us to reconcile pending HTLCs after a restart.
-        db.receiveIncomingPayment(paymentHash, received.amount, received.timestamp)
+        log.info(s"received complete payment for amount=${parts.map(_.amount).sum}")
         pendingPayments.get(paymentHash).foreach {
           case (preimage: ByteVector32, handler: ActorRef) =>
             handler ! PoisonPill
-            parts.foreach(p => commandBuffer ! CommandBuffer.CommandSend(p.payment.fromChannelId, CMD_FULFILL_HTLC(p.htlcId, preimage, commit = true)))
+            doFulfill(ctx, preimage, e)
         }
-        ctx.system.eventStream.publish(received)
         pendingPayments = pendingPayments - paymentHash
-        onSuccess(received)
       }
 
-    case MultiPartPaymentFSM.ExtraHtlcReceived(paymentHash, p, failure) if doHandle(paymentHash) =>
+    case MultiPartPaymentFSM.ExtraPaymentReceived(paymentHash, p, failure) if doHandle(paymentHash) =>
       Logs.withMdc(log)(Logs.mdc(paymentHash_opt = Some(paymentHash))) {
         failure match {
-          case Some(failure) => commandBuffer ! CommandBuffer.CommandSend(p.payment.fromChannelId, CMD_FAIL_HTLC(p.htlcId, Right(failure), commit = true))
-          // NB: this case shouldn't happen unless the sender violated the spec, so it's ok that we take a slightly more
-          // expensive code path by fetching the preimage from DB.
-          case None => db.getIncomingPayment(paymentHash).foreach(record => {
-            commandBuffer ! CommandBuffer.CommandSend(p.payment.fromChannelId, CMD_FULFILL_HTLC(p.htlcId, record.paymentPreimage, commit = true))
-            db.receiveIncomingPayment(paymentHash, p.payment.amount, p.payment.timestamp)
-            ctx.system.eventStream.publish(PaymentReceived(paymentHash, p.payment :: Nil))
-          })
+          case Some(failure) => p match {
+            case p: MultiPartPaymentFSM.HtlcPart => commandBuffer ! CommandBuffer.CommandSend(p.htlc.channelId, CMD_FAIL_HTLC(p.htlc.id, Right(failure), commit = true))
```

Comment:
I'm not sure the `doFulfill`/`doFail` methods are useful if we can't factor that in.

Reply:
I'm not sure, this code has to be there to handle unexpected cases, but that is a spec violation.
I believe most use-cases can simply ignore that technicality.
But we probably need real use-cases to figure that case-by-case.